rllib/env/tests/test_single_agent_episode.py

```diff
@@ -0,0 +1,303 @@
+
+    def test_add_timestep(self):
+        # Set the random seed (otherwise the episode will terminate at
+        # different points in each test run).
+        np.random.seed(0)
+        # Create an empty episode and add initial observations.
+        episode = SingleAgentEpisode()
+        env = gym.make("CartPole-v1")
+        obs, _ = env.reset()
+        episode.add_initial_observation(initial_observation=obs)
+
+        # Sample 100 timesteps and add them to the episode.
+        for i in range(100):
+            action = env.action_space.sample()
+            obs, reward, is_terminated, is_truncated, _ = env.step(action)
+            episode.add_timestep(
+                observation=obs,
+                action=action,
+                reward=reward,
+                is_terminated=is_terminated,
+                is_truncated=is_truncated,
+            )
+            if is_terminated:
+                break
+
+        # Assert that the episode timestep is at 100.
+        self.assertTrue(episode.t == len(episode.observations) - 1 == i + 1)
+        # Assert that `t_started` stayed at zero.
+        self.assertTrue(episode.t_started == 0)
+        # Assert that all lists have the proper lengths.
+        self.assertTrue(
+            len(episode.actions)
+            == len(episode.rewards)
+            == len(episode.observations) - 1
+            == i + 1
+        )
+        # Assert that the flags are set correctly.
+        self.assertTrue(episode.is_terminated == is_terminated)
+        self.assertTrue(episode.is_truncated == is_truncated)
+        self.assertTrue(episode.is_done == is_terminated or is_truncated)
+
+    def test_create_successor(self):
+        # Create an empty episode.
+        episode_1 = SingleAgentEpisode()
+        # Create an environment.
+        env = TestEnv()
+        # Add initial observation.
+        init_obs, _ = env.reset()
+        episode_1.add_initial_observation(initial_observation=init_obs)
+        # Sample 100 steps.
+        for i in range(100):
+            action = i
+            obs, reward, is_terminated, is_truncated, _ = env.step(action)
+            episode_1.add_timestep(
+                observation=obs,
+                action=action,
+                reward=reward,
+                is_terminated=is_terminated,
+                is_truncated=is_truncated,
+            )
+
+        # Assert that the episode has indeed 100 timesteps.
+        self.assertTrue(episode_1.t == 100)
+
+        # Create a successor.
+        episode_2 = episode_1.create_successor()
+        # Assert that it has the same id.
+        self.assertTrue(episode_1.id_ == episode_2.id_)
+        # Assert that the timestep starts at the end of the last episode.
+        self.assertTrue(episode_1.t == episode_2.t == episode_2.t_started)
+        # Assert that the last observation of `episode_1` is the first of
+        # `episode_2`.
+        self.assertTrue(episode_1.observations[-1] == episode_2.observations[0])
```

Comment:
Can we add another two lines here that test, whether if we (hackish'ly) add stuff the the successor's lists (manually), whether it influences the original episode (it shouldn't)?

Reply:
I put this in. Thanks for this idea @sven1977